File: agrigrow-backend/routes/progress.js
// routes/progress.js — Lesson completion and daily tasks
const express = require('express');
const router = express.Router();
const auth = require('../middleware/auth');
const { dbRun, dbGet, dbAll } = require('../db/database');

async function checkAndAwardBadges(userId) {
  const user = await dbGet('SELECT points FROM users WHERE id = ?', [userId]);
  const lessons = await dbAll('SELECT lesson_id FROM completed_lessons WHERE user_id = ?', [userId]);

  if (lessons.length >= 1)
    await dbRun("UPDATE badges SET earned = 1 WHERE user_id = ? AND badge_id = 'b1' AND earned = 0", [userId]);
  if (user.points >= 500)
    await dbRun("UPDATE badges SET earned = 1 WHERE user_id = ? AND badge_id = 'b4' AND earned = 0", [userId]);
  if (lessons.length >= 6)
    await dbRun("UPDATE badges SET earned = 1 WHERE user_id = ? AND badge_id = 'b5' AND earned = 0", [userId]);
}

// GET /api/progress
router.get('/', auth, async (req, res) => {
  try {
    const user = await dbGet('SELECT points, streak FROM users WHERE id = ?', [req.userId]);
    const lessons = await dbAll('SELECT lesson_id FROM completed_lessons WHERE user_id = ?', [req.userId]);
    const badges  = await dbAll('SELECT badge_id, earned FROM badges WHERE user_id = ?', [req.userId]);

    res.json({
      points: user.points,
      streak: user.streak,
      completedLessons: lessons.map(r => r.lesson_id),
      badges: badges.map(r => ({ id: r.badge_id, earned: r.earned === 1 })),
    });
  } catch (err) {
    console.error('Progress get error:', err);
    res.status(500).json({ error: 'Failed to load progress.' });
  }
});

// POST /api/progress/lesson — Mark lesson complete
router.post('/lesson', auth, async (req, res) => {
  try {
    const { lessonId, pts } = req.body;
    if (!lessonId || pts == null) return res.status(400).json({ error: 'lessonId and pts are required.' });

    const already = await dbGet('SELECT id FROM completed_lessons WHERE user_id = ? AND lesson_id = ?', [req.userId, lessonId]);
    if (already) return res.json({ alreadyDone: true });

    await dbRun('INSERT INTO completed_lessons (user_id, lesson_id, pts) VALUES (?, ?, ?)', [req.userId, lessonId, pts]);
    await dbRun('UPDATE users SET points = points + ? WHERE id = ?', [pts, req.userId]);
    await checkAndAwardBadges(req.userId);

    const today = new Date().toISOString().split('T')[0];
    await dbRun("UPDATE today_tasks SET done = 1 WHERE user_id = ? AND task_id = 't1' AND date = ?", [req.userId, today]);

    const user = await dbGet('SELECT points FROM users WHERE id = ?', [req.userId]);
    const badges = await dbAll('SELECT badge_id, earned FROM badges WHERE user_id = ?', [req.userId]);

    res.json({ points: user.points, badges: badges.map(r => ({ id: r.badge_id, earned: r.earned === 1 })) });
  } catch (err) {
    console.error('Complete lesson error:', err);
    res.status(500).json({ error: 'Failed to save lesson progress.' });
  }
});

// POST /api/progress/task — Toggle daily task
router.post('/task', auth, async (req, res) => {
  try {
    const { taskId } = req.body;
    if (!taskId) return res.status(400).json({ error: 'taskId is required.' });

    const today = new Date().toISOString().split('T')[0];
    const TASK_PTS = { t1: 50, t2: 30, t3: 20 };

    const task = await dbGet('SELECT done FROM today_tasks WHERE user_id = ? AND task_id = ? AND date = ?', [req.userId, taskId, today]);
    if (!task) return res.status(404).json({ error: 'Task not found for today.' });

    const nowDone = task.done === 0 ? 1 : 0;
    const ptsDelta = nowDone ? TASK_PTS[taskId] : -(TASK_PTS[taskId]);

    await dbRun('UPDATE today_tasks SET done = ? WHERE user_id = ? AND task_id = ? AND date = ?', [nowDone, req.userId, taskId, today]);
    await dbRun('UPDATE users SET points = MAX(0, points + ?) WHERE id = ?', [ptsDelta, req.userId]);

    const user = await dbGet('SELECT points FROM users WHERE id = ?', [req.userId]);
    res.json({ done: nowDone === 1, points: user.points });
  } catch (err) {
    console.error('Toggle task error:', err);
    res.status(500).json({ error: 'Failed to update task.' });
  }
});

module.exports = router;

File: flask-backend/routes/progress.py
# Progress tracking and daily tasks

from datetime import datetime
from flask import Blueprint, request, jsonify, g
from db.database import db_run, db_get, db_all
from middleware.auth_middleware import require_auth

progress_bp = Blueprint('progress', __name__)


def check_and_award_badges(user_id):
    """Unlock earned badges."""
    user    = db_get('SELECT points FROM users WHERE id = %s', (user_id,))
    lessons = db_all('SELECT lesson_id FROM completed_lessons WHERE user_id = %s', (user_id,))

    if len(lessons) >= 1:
        db_run("UPDATE badges SET earned=1 WHERE user_id=%s AND badge_id='b1' AND earned=0", (user_id,))
    if user['points'] >= 500:
        db_run("UPDATE badges SET earned=1 WHERE user_id=%s AND badge_id='b4' AND earned=0", (user_id,))
    if len(lessons) >= 6:
        db_run("UPDATE badges SET earned=1 WHERE user_id=%s AND badge_id='b5' AND earned=0", (user_id,))


# Get progress
@progress_bp.route('/', methods=['GET'])
@require_auth
def get_progress():
    try:
        user    = db_get('SELECT points, streak FROM users WHERE id = %s', (g.user_id,))
        lessons = db_all('SELECT lesson_id FROM completed_lessons WHERE user_id = %s', (g.user_id,))
        badges  = db_all('SELECT badge_id, earned FROM badges WHERE user_id = %s', (g.user_id,))

        return jsonify({
            'points': user['points'],
            'streak': user['streak'],
            'completedLessons': [r['lesson_id'] for r in lessons],
            'badges': [{'id': r['badge_id'], 'earned': r['earned'] == 1} for r in badges],
        })

    except Exception as e:
        print('Progress get error:', e)
        return jsonify({'error': 'Failed to load progress.'}), 500


# Complete lesson
@progress_bp.route('/lesson', methods=['POST'])
@require_auth
def complete_lesson():
    try:
        data      = request.get_json()
        lesson_id = data.get('lessonId')
        pts       = data.get('pts')

        if not lesson_id or pts is None:
            return jsonify({'error': 'lessonId and pts are required.'}), 400

        already = db_get(
            'SELECT id FROM completed_lessons WHERE user_id = %s AND lesson_id = %s',
            (g.user_id, lesson_id)
        )
        if already:
            return jsonify({'alreadyDone': True})

        db_run('INSERT INTO completed_lessons (user_id, lesson_id, pts) VALUES (%s, %s, %s)',
               (g.user_id, lesson_id, pts))
        db_run('UPDATE users SET points = points + %s WHERE id = %s', (pts, g.user_id))
        check_and_award_badges(g.user_id)

        today = datetime.now().strftime('%Y-%m-%d')
        db_run("UPDATE today_tasks SET done=1 WHERE user_id=%s AND task_id='t1' AND date=%s",
               (g.user_id, today))

        user   = db_get('SELECT points FROM users WHERE id = %s', (g.user_id,))
        badges = db_all('SELECT badge_id, earned FROM badges WHERE user_id = %s', (g.user_id,))

        return jsonify({
            'points': user['points'],
            'badges': [{'id': r['badge_id'], 'earned': r['earned'] == 1} for r in badges],
        })

    except Exception as e:
        print('Complete lesson error:', e)
        return jsonify({'error': 'Failed to save lesson progress.'}), 500


# Toggle task
@progress_bp.route('/task', methods=['POST'])
@require_auth
def toggle_task():
    try:
        data    = request.get_json()
        task_id = data.get('taskId')

        if not task_id:
            return jsonify({'error': 'taskId is required.'}), 400

        today    = datetime.now().strftime('%Y-%m-%d')
        TASK_PTS = {'t1': 50, 't2': 30, 't3': 20}

        task = db_get(
            'SELECT done FROM today_tasks WHERE user_id=%s AND task_id=%s AND date=%s',
            (g.user_id, task_id, today)
        )
        if not task:
            return jsonify({'error': 'Task not found for today.'}), 404

        now_done  = 0 if task['done'] == 1 else 1
        pts_delta = TASK_PTS.get(task_id, 0) if now_done else -TASK_PTS.get(task_id, 0)

        db_run('UPDATE today_tasks SET done=%s WHERE user_id=%s AND task_id=%s AND date=%s',
               (now_done, g.user_id, task_id, today))
        db_run('UPDATE users SET points = GREATEST(0, points + %s) WHERE id = %s',
               (pts_delta, g.user_id))

        user = db_get('SELECT points FROM users WHERE id = %s', (g.user_id,))
        return jsonify({'done': now_done == 1, 'points': user['points']})

    except Exception as e:
        print('Toggle task error:', e)
        return jsonify({'error': 'Failed to update task.'}), 500
